// themeable/build/resolveTheme.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.resolveTheme = resolveTheme;
exports.default = void 0;

var _isTheme = _interopRequireDefault(require("./isTheme"));

var _isThemeable = _interopRequireDefault(require("./isThemeable"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Licensed under the Apache License, Version 2.0 (the “License”); you may not
// use this file except in compliance with the License. You may obtain a copy of
// the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an “AS IS” BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.
function _resolveTheme(source
/*:string*/
, data
/*:Object*/
, identifier
/*:string*/
) {
  if (!data) return undefined;
  var theme = data.theme;
  if (!theme) return undefined;

  if ((0, _isTheme.default)(theme)) {
    return !identifier ? undefined : {
      componentTheme: theme.componentThemes[identifier],
      identifier: identifier,
      source: source,
      theme: theme
    };
  }

  return {
    componentTheme: theme,
    identifier: undefined,
    source: source,
    theme: undefined
  };
}
/*:: export type ResolvedTheme = {
  componentTheme:ComponentTheme,
  identifier?:string,
  source:string,
  theme?:Theme
}*/


/**
 * Tries to resolve the component theme from the properties and context.
 */
function resolveTheme(target
/*:Component<*>*/
)
/*:ResolvedTheme*/
{
  if (!(0, _isThemeable.default)(target)) return undefined;
  var context = target.context,
      props = target.props;
  var IDENTIFIER = target.getThemeableIdentifier();
  return _resolveTheme('property', props, IDENTIFIER) || _resolveTheme('context', context, IDENTIFIER) || undefined;
}

var _default = resolveTheme;
exports.default = _default;
